### PR TITLE
Add a feature for private room and add a feature for invitation Close #13 #14

### LINKED ISSUE #13: Add a feature for private chat rooms
## Requirements
- Users can create private chat rooms which non-member users cannot see in the room list
- Owner users can invite non-member users to the room regardless of the publicity of the room
  - In other words, users cannot join in private rooms without invitation

### PR DESCRIPTION
**Pull request by CaptainYANG**
#13 #14

### CODE REVIEW

File: client/InRoom.html [hunk 1]
```diff
@@ -1,8 +1,11 @@
 <template name="inRoom">
-    {{#if currentRoom}}
+    {{#with currentRoom}}
```
Comment (exKAZUu) [inline client/InRoom.html]: :+1:

File: client/InRoom.js [hunk 2]
```diff
@@ -3,17 +3,14 @@
...
 Template.inRoom.helpers({
     currentRoom: function () {
-        const room = Rooms.findOne({_id: FlowRouter.getParam("roomId")});
-        if(room!=null){
-            return room.roomName;
-        }else{
-            FlowRouter.go('notFound');
-            return false;
-        }
+        return Rooms.findOne({_id: FlowRouter.getParam("roomId")});
```
Comment (exKAZUu) [inline client/InRoom.js]: You can write: `return Rooms.findOne(FlowRouter.getParam("roomId"));`

File: client/Invite.js [hunk 3]
```diff
@@ -0,0 +1,24 @@
+/**
+ * Created by yuyin on 2016/3/8.
+ */
+Template.invite.helpers({
+    users:function(){
+        var room = Rooms.findOne({_id:FlowRouter.getParam("roomId")});
```
Comment (exKAZUu) [inline client/Invite.js]: You can rewrite this as follows: `var room = Rooms.findOne(FlowRouter.getParam("roomId"));`

File: client/Invite.js [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+        var members = room.members;
+        return Meteor.users.find({_id: {$nin: members}});
```
Comment (exKAZUu) [inline client/Invite.js]: You can rewrite this as follows: `return Meteor.users.find({$nin: members});`

File: collections/Rooms.js [hunk 5]
```diff
@@ -19,6 +19,12 @@ RoomSchema = new SimpleSchema({
         autoValue:function(){
             return Meteor.userId();
         }
+    },
+    private:{
+        type:Boolean,
+    },
+    members:{
```
Comment (exKAZUu) [inline collections/Rooms.js]: I'd suggest you to rename this to `memberIds`.

File: client/CreateRoom.js [hunk 6]
```diff
@@ -16,7 +31,19 @@ Template.createRoom.events({
...
         event.preventDefault();
         var text = event.target.text.value;
-        Meteor.call("addRoom", text,function(error,result){
+        var privateRoom = document.querySelector('input[name=access]:checked').value;
+        var selectedMembers=document.querySelector('select').selectedOptions;
+        var members= new Array();
+        for(var i=0;i<selectedMembers.length;i++){
+            members[i]=selectedMembers[i].value;
+        }
+        members.push(Meteor.userId());
```
Comment (exKAZUu) [inline client/CreateRoom.js]: You can rewrite the following code
```
        var members= new Array();
        for(var i=0;i<selectedMembers.length;i++){
            members[i]=selectedMembers[i].value;
        }
```
as follows:
```
        var members = _.map(selectedMembers, function(selectedMember) {
            return selectedMember.value;
        });
```

File: client/CreateRoom.js [hunk 7]
```diff
@@ -2,11 +2,26 @@
...
     "click .confirm": function () {
-        Meteor.call("addRoom", document.getElementById("roomName").value,function(error,result){
+        //todo:clean this code
+        var privateRoom = document.querySelector('input[name=access]:checked').value;
+        var selectedMembers=document.querySelector('select').selectedOptions;
+        var members= new Array();
+        for(var i=0;i<selectedMembers.length;i++){
+            members[i]=selectedMembers[i].value;
+        }
+        members.push(Meteor.userId());
```
Comment (exKAZUu) [inline client/CreateRoom.js]: You can rewrite the following code
```
        var members= new Array();
        for(var i=0;i<selectedMembers.length;i++){
            members[i]=selectedMembers[i].value;
        }
```
as follows:
```
        var members = _.map(selectedMembers, function(selectedMember) {
            return selectedMember.value;
        });
```

File: client/CreateRoom.js [hunk 8]
```diff
@@ -2,11 +2,26 @@
...
+        if(privateRoom=='private'){
+            privateRoom = true;
+        }else{
+            privateRoom = false;
+        }
```
Comment (exKAZUu) [inline client/CreateRoom.js]: You can rewrite this line as follows: `privateRoom = privateRoom === 'private';`

File: client/Invite.js [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+    }
+});
+Template.invite.events({
+    "click .confirm": function () {
+        var selectedMembers=document.querySelector('select').selectedOptions;
+        var members= new Array();
+        for(var i=0;i<selectedMembers.length;i++){
+            members[i]=selectedMembers[i].value;
+        }
+        Meteor.call("inviteMembers", FlowRouter.getParam("roomId"),members);
```
Comment (exKAZUu) [inline client/Invite.js]: You can rewrite this loop using `_.map`

File: client/CreateRoom.js (hunk 6 above)
Comment (exKAZUu) [inline client/CreateRoom.js]: Please see http://underscorejs.org/#map

File: client/Invite.js (hunk 4 above)
Comment (CaptainYANG) [inline client/Invite.js]: It will occur error because of the logic operator.

Comment (exKAZUu) [inline client/Invite.js]: I see. Sorry for that.

**Comment by ss0201** [comment]
LGTM

**Pull request merged**